File: lab4-records/haskell/uniontest.hs
module Main where

data Month -- enum type definition
    = JAN | FEB | MAR | APR | MAY | JUN
    | JUL | AUG | SEP | OCT | NOV | DEC
    deriving (Show)

type Day_Number = Int
type Year = Int

data Date -- defining union record type
    = DateDMYConstr -- this is a value constructor name
    {
        date_d :: Day_Number,
        date_m :: Month,
        date_y :: Year
    }
    | DateDYConstr -- alternative value constructor
    {
        date_d :: Day_Number,
        date_y :: Year    
    }

instance (Show  Date) where
    show (DateDMYConstr d m y) =
        (show d) ++ "/" ++
        (show m) ++ "/" ++
        (show y)
    show (DateDYConstr d y) =
        (show d) ++ "//" ++
        (show y)

theDay = DateDMYConstr { date_d = 1, date_m = JAN, date_y = 1901 }
theDay2 = DateDYConstr 111 1901

main =
    do
    putStrLn $ show $ theDay
    putStrLn $ show $ theDay2
    putStrLn $ show $ theDay2 { date_d = 1 + (date_d theDay2) }

File: lab4-records/java/UnionTest.java
public class UnionTest
{

    enum Month {JAN, FEB, MAR, APR, MAY, JUN, 
                JUL, AUG, SEP, OCT, NOV, DEC};

    public static abstract class Date
    {
        public int year;
        // not enough data -- therefore abstract
        
        public Date(int y) { year = y; }
    }
    
    public static class DateDMY extends Date
    {
        public int mday;
        public Month month;
        
        public DateDMY(int d, Month m, int y) 
        {
            super(y);
            mday = d;
            month = m;
        }
        
        public String toString()
        {
            return "" + mday + "/" + month + "/" + year;
        }
    }
    
    public static class DateDY extends Date
    {
        public int yday;
        
        public DateDY(int d, int y) 
        {
            super(y);
            yday = d;
        }
        
        public String toString()
        {
            return "" + yday + "//" + year;
        }
    }
    
    
    
    public static void main(String[] args)
    {
        Date someday;
    
        someday = new DateDMY(1,Month.JAN,1901);
        System.out.println(someday);
        
        someday = new DateDY(111,1901);
        System.out.println(someday);
        
        if ( someday instanceof DateDMY )
        {
            System.out.println("someday is a DMY");
        }
        else if ( someday instanceof DateDY )
        {
            System.out.println("someday is a DY");
        }
        else
        {
            System.out.println("someday is neither DY nor DMY");
        }
        
    }    
}